test/unit/app/controllers/permissions/permissions-controller-test.js

```diff
@@ -0,0 +1,1191 @@
+import assert from 'assert'
+import { find } from 'lodash'
+
+import {
+  METADATA_STORE_KEY,
+  WALLET_PREFIX,
+} from '../../../../../app/scripts/controllers/permissions/enums'
+
+import getRestrictedMethods
+  from '../../../../../app/scripts/controllers/permissions/restrictedMethods'
+
+import {
+  addInternalMethodPrefix,
+} from '../../../../../app/scripts/controllers/permissions'
+
+import {
+  getPermController,
+  grantPermissions,
+  getNotifyDomain,
+  getNotifyAllDomains,
+  getApprovedPermissionsRequest,
+  ACCOUNT_ARRAYS,
+  DUMMY_ACCOUNT,
+  ERRORS,
+  ORIGINS,
+  PERM_NAMES,
+  PERMS,
+  NOTIFICATIONS,
+  REQUEST_IDS,
+} from './mocks'
+
+let permController
```

Comment:
This mock is being shared between all of the tests, making them interdependent. It would be better to initialize it anew in each test (e.g. `const permController = getPermController()`. It's easy for tests to become accidentally dependant on each other when mocks are shared.

The same goes for the notifications as well, and any other shared state in the mocks.

Reply:
Variables in the module context shared between tests are a sign that something is amiss, as they break test isolation. Though I sometimes make exceptions for things instantiated by `sinon`, as they can be reliably reset to their initial states by calling `sinon.restore()` in the `afterEach` hook ([as demonstrated here](https://sinonjs.org/releases/v9.0.0/general-setup/))

Generally my approach to testing create a shared set of stubs/spies that all of the tests can access, but construct the object under test within each individual test (passing in the stubs/spies).

Reply:
Resolved by instantiating all constants in each file. Some "constants" e.g. `NOTIFICATIONS` were actually an object of getter functions. All such "constants" have been moved to the export `getters` in `mocks`

Reply:
This mock is still being shared between tests :confused: I'm not sure what you mean about the constants, or how that relates to this comment.

Reply:
Resolved by ensuring that variables shared over multiple tests are either immutable or re-initialized (using `beforeEach`).